server/src/actors/project.rs

```diff
@@ -36,38 +51,57 @@ pub enum ProjectError {
 
 impl ResponseError for ProjectError {}
 
+#[derive(Clone, Copy, Debug)]
+enum StateSource {
+    FromUpstream,
+    FromDisk,
```

Comment:
_nit_: Ideally, these would just be `Upstream` and `Disk`, see [clippy::enum_variant_names](https://rust-lang-nursery.github.io/rust-clippy/current/index.html#enum_variant_names) (currently disabled for us, though).